nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-security/src/main/java/org/apache/nifi/web/security/jwt/JwtAuthenticationFilter.java

```diff
@@ -43,33 +51,63 @@ public Authentication attemptAuthentication(final HttpServletRequest request) {
             return null;
         }
 
-        // TODO: Refactor request header extraction logic to shared utility as it is duplicated in AccessResource
+        // Check for JWT in cookie and header
+        final String cookieToken = getTokenFromCookie(request);
+        final String headerToken = getTokenFromHeader(request);
+        if (cookieToken != null && !cookieToken.isEmpty()) {
+            if (!UNAUTHENTICATED_METHODS.contains(request.getMethod().toUpperCase())) {
+                // To protect against CSRF when using a cookie, if the request method requires authentication the request must have a matching Authorization header JWT
+                if (headerToken.equals(cookieToken)) {
+                    return new JwtAuthenticationRequestToken(headerToken, request.getRemoteAddr());
+                } else {
+                    throw new InvalidAuthenticationException("Authorization HTTP header and authentication cookie did not match.");
+                }
+            } else {
+                return new JwtAuthenticationRequestToken(cookieToken, request.getRemoteAddr());
+            }
+        } else if (headerToken != null && !headerToken.isEmpty()) {
+            return new JwtAuthenticationRequestToken(headerToken, request.getRemoteAddr());
+        } else {
+            return null;
+        }
+    }
+
+    private boolean validJwtFormat(String authenticationHeader) {
+        Matcher matcher = BEARER_HEADER_PATTERN.matcher(authenticationHeader);
+        return matcher.matches();
+    }
 
-        // get the principal out of the user token
+    private String getTokenFromHeader(final HttpServletRequest request) {
         final String authorizationHeader = request.getHeader(AUTHORIZATION);
 
         // if there is no authorization header, we don't know the user
         if (authorizationHeader == null || !validJwtFormat(authorizationHeader)) {
             return null;
         } else {
             // Extract the Base64 encoded token from the Authorization header
-            final String token = getTokenFromHeader(authorizationHeader);
-            return new JwtAuthenticationRequestToken(token, request.getRemoteAddr());
+            return getTokenFromHeader(authorizationHeader);
         }
     }
 
-    private boolean validJwtFormat(String authenticationHeader) {
-        Matcher matcher = tokenPattern.matcher(authenticationHeader);
-        return matcher.matches();
-    }
-
     public static String getTokenFromHeader(String authenticationHeader) {
-        Matcher matcher = tokenPattern.matcher(authenticationHeader);
+        Matcher matcher = BEARER_HEADER_PATTERN.matcher(authenticationHeader);
         if(matcher.matches()) {
             return matcher.group(1);
         } else {
             throw new InvalidAuthenticationException("JWT did not match expected pattern.");
         }
     }
 
+    private String getTokenFromCookie(final HttpServletRequest request) {
```

Comment:
Spring `WebUtils` provides a `getCookie` method that could be used to simplify this method.